### PR TITLE
8343148: C2: Refactor uses of "PhaseValue::*con*() + PhaseIdealLoop::set_ctrl()" into separate method

### PR DESCRIPTION
**Pull request by theoweidmannoracle**
This patch introduces the methods `PhaseIdealLoop::intcon` and `PhaseIdealLoop::longcon` which are wrappers for:

```
ConINode* node = _igvn.intcon(i);
set_ctrl(node, C->root());
```

and

```
ConLNode* node = _igvn.longcon(i);
set_ctrl(node, C->root());
```

Occurrences of this pattern in loopnode.cpp were replaced with the appropriate call to the new methods.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343148](https://bugs.openjdk.org/browse/JDK-8343148): C2: Refactor uses of "PhaseValue::*con*() + PhaseIdealLoop::set_ctrl()" into separate method (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21836/head:pull/21836` \
`$ git checkout pull/21836`

Update a local copy of the PR: \
`$ git checkout pull/21836` \
`$ git pull https://git.openjdk.org/jdk.git pull/21836/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21836`

View PR using the GUI difftool: \
`$ git pr show -t 21836`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21836.diff">https://git.openjdk.org/jdk/pull/21836.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21836#issuecomment-2457160496)
</details>

### CODE REVIEW

**Comment by chhagedorn** [comment]
While at it, we could extend this to the other constant creation methods `zerocon`, `makecon`, and `integercon` as well (`uncached_makecon` is only called by the other `*con*` methods - could be made `private` at some point). I suggest to update the RFE title accordingly since it only mentions `intcon` now. Maybe something like `PhaseValue::*con*() + ...`.

**Review by vnkozlov — COMMENTED** [review]
Do we have other places (not new constant node) where we set Root as control?
May be we can add `set_root_as_ctrl(n)` method in `loop node.hpp` in such case.

**Comment by vnkozlov** [comment]
> Do we have other places (not new constant node) where we set Root as control? 

In loop opts.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -2972,8 +2960,7 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
 
       // Kill the eliminated test
       C->set_major_progress();
-      Node *kill_con = _igvn.intcon(1-flip);
-      set_ctrl(kill_con, C->root());
+      Node *kill_con = intcon(1-flip);
```
Comment (eme64) [inline src/hotspot/share/opto/loopTransform.cpp:2963]:
```suggestion
      Node* kill_con = intcon(1-flip);
```
We generally now have the pointer `*` with the type. So if you touch any new code please update it ;)

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -331,8 +331,7 @@ void PhaseIdealLoop::dominated_by(IfProjNode* prevdom, IfNode* iff, bool flip, b
       pop = Op_IfTrue;
   }
   // 'con' is set to true or false to kill the dominated test.
-  Node *con = _igvn.makecon(pop == Op_IfTrue ? TypeInt::ONE : TypeInt::ZERO);
-  set_ctrl(con, C->root()); // Constant gets a new use
+  Node *con = makecon(pop == Op_IfTrue ? TypeInt::ONE : TypeInt::ZERO);
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:334]:
```suggestion
  Node* con = makecon(pop == Op_IfTrue ? TypeInt::ONE : TypeInt::ZERO);
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 3]
```diff
@@ -2906,8 +2904,7 @@ Node* PhaseIdealLoop::short_circuit_if(IfNode* iff, ProjNode* live_proj) {
   guarantee(live_proj != nullptr, "null projection");
   int proj_con = live_proj->_con;
   assert(proj_con == 0 || proj_con == 1, "false or true projection");
-  Node *con = _igvn.intcon(proj_con);
-  set_ctrl(con, C->root());
+  Node *con = intcon(proj_con);
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:2907]:
```suggestion
  Node* con = intcon(proj_con);
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 4]
```diff
@@ -3246,8 +3242,7 @@ void PhaseIdealLoop::remove_cmpi_loop_exit(IfNode* if_cmp, IdealLoopTree *loop)
   assert(if_cmp->in(1)->in(1)->Opcode() == Op_CmpI &&
          stay_in_loop(lp_proj, loop)->is_If() &&
          stay_in_loop(lp_proj, loop)->in(1)->in(1)->Opcode() == Op_CmpU, "inserted cmpi before cmpu");
-  Node *con = _igvn.makecon(lp_proj->is_IfTrue() ? TypeInt::ONE : TypeInt::ZERO);
-  set_ctrl(con, C->root());
+  Node *con = makecon(lp_proj->is_IfTrue() ? TypeInt::ONE : TypeInt::ZERO);
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:3245]:
```suggestion
  Node* con = makecon(lp_proj->is_IfTrue() ? TypeInt::ONE : TypeInt::ZERO);
```

**Comment by theoweidmannoracle** [comment]
> Do we have other places (not new constant node) where we set Root as control? May be we can add `set_root_as_ctrl(n)` method in `loop node.hpp` in such case.

There's only three locations where control is set to the root in the loop files now (not counting the ones in the new methods I added). The main reason for this patch is bugs caused by people forgetting to set control for constants (e.g. https://bugs.openjdk.org/browse/JDK-8343137), which is now prevented if the new helper methods are used.

Do you think there would be any benefit from introducing `set_root_as_ctrl(n)` given there's only about three places where this pattern occurs now?

**Comment by vnkozlov** [comment]
> > Do we have other places (not new constant node) where we set Root as control? May be we can add `set_root_as_ctrl(n)` method in `loop node.hpp` in such case.
> 
> There's only three locations where control is set to the root in the loop files now (not counting the ones in the new methods I added). The main reason for this patch is bugs caused by people forgetting to set control for constants (e.g. https://bugs.openjdk.org/browse/JDK-8343137), which is now prevented if the new helper methods are used.
> 
> Do you think there would be any benefit from introducing `set_root_as_ctrl(n)` given there's only about three places where this pattern occurs now?

My suggesting is about additional cleaning code. I think 3 + 5 places are enough to justify to have a new function in header file. Also `set_root_as_ctrl(n)` could be copy if `set_ctrl(n, ctrl)` without 2 asserts which checks `ctrl`. It will be faster.

**Comment by theoweidmannoracle** [comment]
@vnkozlov I implemented your suggestion. Would you like to take another look? (It also helped me discover some more cases which can be replaced with the new *con*() functions.)

File: src/hotspot/share/opto/loopnode.hpp [hunk 5]
```diff
@@ -992,6 +992,10 @@ class PhaseIdealLoop : public PhaseTransform {
     assert( ctrl == find_non_split_ctrl(ctrl), "must set legal crtl" );
     _loop_or_ctrl.map(n->_idx, (Node*)((intptr_t)ctrl + 1));
   }
+  void set_root_as_ctrl(Node* n) {
+    assert( !has_node(n) || has_ctrl(n), "" );
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.hpp:996]: We don't use spaces after and before `()` in assert(). Ignore old style in previous lines.

File: src/hotspot/share/opto/loopnode.cpp [hunk 6]
```diff
@@ -3155,7 +3144,7 @@ void OuterStripMinedLoopNode::transform_to_counted_loop(PhaseIterGVN* igvn, Phas
   // make counted loop exit test always fail
   ConINode* zero = igvn->intcon(0);
   if (iloop != nullptr) {
-    iloop->set_ctrl(zero, igvn->C->root());
+    iloop->set_root_as_ctrl(zero);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:3147]: Please look on history of this code. This is suspicious - constant nodes should be always attached to Root.

File: src/hotspot/share/opto/loopopts.cpp [hunk 7]
```diff
@@ -190,7 +190,7 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
 
     if (x->is_Con()) {
       // Constant's control is always root.
-      set_ctrl(x, C->root());
+      set_root_as_ctrl(x);
       continue;
     }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopopts.cpp:195]: This looks like "band-aid" - this should be assert. May be investigate in separate RFE.

**Review by vnkozlov — COMMENTED** [review]
Looks good. I have few additional comments.

File: src/hotspot/share/opto/loopopts.cpp (hunk 7 above)
Comment (theoweidmannoracle) [reply src/hotspot/share/opto/loopopts.cpp:195]: I opened an RFE for this https://bugs.openjdk.org/browse/JDK-8343907

Comment (chhagedorn) [reply src/hotspot/share/opto/loopopts.cpp:195]: If you modify the following code above to use your new `makecon()` (could be done either way), could this then be turned into an assert? By looking at the code, it suggests that we only miss to set ctrl in the `singleton` case which would then be covered.
https://github.com/openjdk/jdk/blob/5ca6698ba418e82ff93471fbb495759850f26f63/src/hotspot/share/opto/loopopts.cpp#L123-L125
You could also only change `makecon()` above and revisit this code later again to remove the `set_root_as_ctrl()` and add an assert.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 8]
```diff
@@ -2059,8 +2051,7 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
   // Kill the clone's backedge
   Node *newcle = old_new[loop_end->_idx];
   _igvn.hash_delete(newcle);
-  Node *one = _igvn.intcon(1);
-  set_ctrl(one, C->root());
+  Node *one = intcon(1);
```
[the remark this answers is not in the document]
Comment (theoweidmannoracle) [reply src/hotspot/share/opto/loopTransform.cpp:2054]: Thanks, Christian, I have now configured my idea to take care of fixing the formatting of changed lines.

File: src/hotspot/share/opto/loopopts.cpp (hunk 7 above)
Comment (theoweidmannoracle) [reply src/hotspot/share/opto/loopopts.cpp]: I added the asserts as @chhagedorn suggested. All tests in the internal testing passed with this new assert.

File: src/hotspot/share/opto/loopopts.cpp [hunk 9]
```diff
@@ -189,9 +189,7 @@ Node* PhaseIdealLoop::split_thru_phi(Node* n, Node* region, int policy) {
     IdealLoopTree *old_loop;
 
     if (x->is_Con()) {
-      // Constant's control is always root.
-      set_root_as_ctrl(x);
-      continue;
+      assert(get_ctrl(x) == C->root(), "constant control is not root");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:192]: I think you should still execute `continue` here because we want to skip constants.

**Review by chhagedorn — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/loopopts.cpp (hunk 9 above)
Comment (theoweidmannoracle) [reply src/hotspot/share/opto/loopopts.cpp:192]: Oops, that slipped through. Thanks! I fixed it.

File: src/hotspot/share/opto/loopnode.cpp [hunk 10]
```diff
@@ -6866,6 +6853,36 @@ void PhaseIdealLoop::rpo(Node* start, Node_Stack &stk, VectorSet &visited, Node_
   }
 }
 
+ConINode* PhaseIdealLoop::intcon(jint i) {
+  ConINode* node = _igvn.intcon(i);
+  set_root_as_ctrl(node);
+  return node;
+}
+
+ConLNode *PhaseIdealLoop::longcon(jlong i) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:6862]:
```suggestion
ConLNode* PhaseIdealLoop::longcon(jlong i) {
```

File: src/hotspot/share/opto/loopnode.cpp [hunk 11]
```diff
@@ -6866,6 +6853,36 @@ void PhaseIdealLoop::rpo(Node* start, Node_Stack &stk, VectorSet &visited, Node_
...
+  ConLNode* node = _igvn.longcon(i);
+  set_root_as_ctrl(node);
+  return node;
+}
+
+ConNode *PhaseIdealLoop::makecon(const Type* t) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:6868]:
```suggestion
ConNode* PhaseIdealLoop::makecon(const Type* t) {
```

File: src/hotspot/share/opto/loopnode.cpp [hunk 12]
```diff
@@ -6866,6 +6853,36 @@ void PhaseIdealLoop::rpo(Node* start, Node_Stack &stk, VectorSet &visited, Node_
...
+  return node;
+}
+
+ConNode* PhaseIdealLoop::integercon(jlong l, BasicType bt) {
+  ConNode* node = _igvn.integercon(l, bt);
+  set_root_as_ctrl(node);
+  return node;
+}
+
+ConNode *PhaseIdealLoop::zerocon(BasicType bt) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:6880]:
```suggestion
ConNode* PhaseIdealLoop::zerocon(BasicType bt) {
```

**Review by chhagedorn — APPROVED** [review]
Some last minor comments, otherwise the updates look good to me. Thanks for doing this extended refactoring!

**Review by chhagedorn — APPROVED** [review]

**Comment by theoweidmannoracle** [comment]
@vnkozlov Could you take another look at the latest changes?

**Review by vnkozlov — APPROVED** [review]
This version looks good.

**Pull request closed without merging**